### PR TITLE
Add broadcast table support

### PR DESCRIPTION
**Pull request by Novemser**
Using statistic information for table broadcast support.

### CODE REVIEW

**Comment by Novemser** [comment]
Micro benchmark
```
TPCH 1G
Before:
q2 -x5 elapsed time: 46080560699ns
With broadcast:
q2 -x5 elapsed time: 35764860785ns
```
Around 22.4% speed up.

Broadcast plan:
```
q2:
== Physical Plan ==
*Sort [s_acctbal#53 DESC NULLS LAST, n_name#21 ASC NULLS FIRST, s_name#49 ASC NULLS FIRST, p_partkey#118L ASC NULLS FIRST], true, 0
+- Exchange rangepartitioning(s_acctbal#53 DESC NULLS LAST, n_name#21 ASC NULLS FIRST, s_name#49 ASC NULLS FIRST, p_partkey#118L ASC NULLS FIRST, 200)
   +- *Project [s_acctbal#53, s_name#49, n_name#21, p_partkey#118L, p_mfgr#120, s_address#50, s_phone#52, s_comment#54]
      +- *BroadcastHashJoin [n_regionkey#22L], [r_regionkey#12L], Inner, BuildRight
         :- *Project [p_partkey#118L, p_mfgr#120, s_name#49, s_address#50, s_phone#52, s_acctbal#53, s_comment#54, n_name#21, n_regionkey#22L]
         :  +- *BroadcastHashJoin [s_nationkey#51L], [n_nationkey#20L], Inner, BuildRight
         :     :- *Project [p_partkey#118L, p_mfgr#120, s_name#49, s_address#50, s_nationkey#51L, s_phone#52, s_acctbal#53, s_comment#54]
         :     :  +- *BroadcastHashJoin [ps_suppkey#1L], [s_suppkey#48L], Inner, BuildRight
         :     :     :- *Project [p_partkey#118L, p_mfgr#120, ps_suppkey#1L]
         :     :     :  +- *SortMergeJoin [ps_supplycost#3, p_partkey#118L], [min(ps_supplycost)#153, ps_partkey#0L#154L], Inner
         :     :     :     :- *Sort [ps_supplycost#3 ASC NULLS FIRST, p_partkey#118L ASC NULLS FIRST], false, 0
         :     :     :     :  +- Exchange hashpartitioning(ps_supplycost#3, p_partkey#118L, 200)
         :     :     :     :     +- *Project [p_partkey#118L, p_mfgr#120, ps_suppkey#1L, ps_supplycost#3]
         :     :     :     :        +- *BroadcastHashJoin [p_partkey#118L], [ps_partkey#0L], Inner, BuildLeft
         :     :     :     :           :- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, false]))
         :     :     :     :           :  +- *Project [p_partkey#118L, p_mfgr#120]
         :     :     :     :           :     +- *Filter EndsWith(p_type#122, BRASS)
         :     :     :     :           :        +- TiDB CoprocessorRDD{[table: part] , Columns: [p_partkey], [p_mfgr], [p_size], [p_type], Filter: [[p_size] EQUAL 15]}
         :     :     :     :           +- TiDB CoprocessorRDD{[table: partsupp] , Columns: [ps_partkey], [ps_suppkey], [ps_supplycost]}
         :     :     :     +- *Sort [min(ps_supplycost)#153 ASC NULLS FIRST, ps_partkey#0L#154L ASC NULLS FIRST], false, 0
         :     :     :        +- Exchange hashpartitioning(min(ps_supplycost)#153, ps_partkey#0L#154L, 200)
         :     :     :           +- *Filter isnotnull(min(ps_supplycost)#153)
         :     :     :              +- *HashAggregate(keys=[ps_partkey#0L], functions=[min(ps_supplycost#3)])
         :     :     :                 +- Exchange hashpartitioning(ps_partkey#0L, 200)
         :     :     :                    +- *HashAggregate(keys=[ps_partkey#0L], functions=[partial_min(ps_supplycost#3)])
         :     :     :                       +- *Project [ps_partkey#0L, ps_supplycost#3]
         :     :     :                          +- *BroadcastHashJoin [n_regionkey#22L], [r_regionkey#12L], Inner, BuildRight
         :     :     :                             :- *Project [ps_partkey#0L, ps_supplycost#3, n_regionkey#22L]
         :     :     :                             :  +- *BroadcastHashJoin [s_nationkey#51L], [n_nationkey#20L], Inner, BuildRight
         :     :     :                             :     :- *Project [ps_partkey#0L, ps_supplycost#3, s_nationkey#51L]
         :     :     :                             :     :  +- *BroadcastHashJoin [ps_suppkey#1L], [s_suppkey#48L], Inner, BuildRight
         :     :     :                             :     :     :- TiDB CoprocessorRDD{[table: partsupp] , Columns: [ps_partkey], [ps_suppkey], [ps_supplycost]}
         :     :     :                             :     :     +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, false]))
         :     :     :                             :     :        +- TiDB CoprocessorRDD{[table: supplier] , Columns: [s_suppkey], [s_nationkey]}
         :     :     :                             :     +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, false]))
         :     :     :                             :        +- TiDB CoprocessorRDD{[table: nation] , Columns: [n_nationkey], [n_regionkey]}
         :     :     :                             +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, false]))
         :     :     :                                +- *Project [r_regionkey#12L]
         :     :     :                                   +- TiDB CoprocessorRDD{[table: region] , Columns: [r_regionkey], [r_name], Filter: [[r_name] EQUAL "EUROPE"]}
         :     :     +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, false]))
         :     :        +- TiDB CoprocessorRDD{[table: supplier] , Columns: [s_suppkey], [s_name], [s_address], [s_nationkey], [s_phone], [s_acctbal], [s_comment]}
         :     +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, false]))
         :        +- TiDB CoprocessorRDD{[table: nation] , Columns: [n_nationkey], [n_name], [n_regionkey]}
         +- BroadcastExchange HashedRelationBroadcastMode(List(input[0, bigint, false]))
            +- *Project [r_regionkey#12L]
               +- TiDB CoprocessorRDD{[table: region] , Columns: [r_regionkey], [r_name], Filter: [[r_name] EQUAL "EUROPE"]}

```

File: core/src/main/scala/org/apache/spark/sql/TiContext.scala [hunk 1]
```diff
@@ -143,9 +143,17 @@ class TiContext(val session: SparkSession) extends Serializable with Logging {
...
+        statisticsManager.tableStatsFromStorage(table)
+      }
+
       val rel: TiDBRelation = new TiDBRelation(
         tiSession,
-        new TiTableReference(dbName, table.getName),
+        new TiTableReference(
+          dbName,
+          table.getName,
+          64 * statisticsManager.getTableCount(table.getId)
```
Comment (Novemser) [inline core/src/main/scala/org/apache/spark/sql/TiContext.scala]: Magic number here, we may do some more precise calculation here later.

**Comment by Novemser** [comment]
@birdstorm @ilovesoup PTAL

File: core/src/main/scala/org/apache/spark/sql/TiContext.scala [hunk 2]
```diff
@@ -143,9 +143,16 @@ class TiContext(val session: SparkSession) extends Serializable with Logging {
       db <- meta.getDatabase(dbName)
       table <- meta.getTables(db)
     } {
+      var sizeInBytes = Long.MaxValue
+      if (loadStatistics) {
+        statisticsManager.tableStatsFromStorage(table)
+        sizeInBytes = 64 * statisticsManager.getTableCount(table.getId)
```
Comment (birdstorm) [inline core/src/main/scala/org/apache/spark/sql/TiContext.scala]: I think we can change this logic to solve overflow
```
val count = statisticManager.getTableCount(table.getId)
if (Long.MaxValue / count < 64) sizeInBytes = 64 * count
```

Comment (birdstorm) [reply core/src/main/scala/org/apache/spark/sql/TiContext.scala]: if we use `sizeInBytes < 0` as condition, chances are that the overflow Long value is a positive arbitrary number

**Pull request merged**